(CDAP-12423) Fix program state and scheduler

chtyim opened the pull request:
Changes are organized by commits.

chtyim committed:
(CDAP-12423) Fix program status publish and consumption.

- Class refactoring
  - Allow batch processing of notifications
  - Allow partial processing of a batch
  - Added metrics for progress tracking
  - Added trace log
  - Use transaction to fetch data partition events

- Correct program state publishing
  - Avoid publishing duplicate completion states from the event handler.
  - Implement StateChangeListener.init to publish states in correct order.

- Fix flaky tests

chtyim commented:
Building https://builds.cask.co/browse/CDAP-RUT1279

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/common/twill/TwillAppLifecycleEventHandler.java:
```diff
@@ -59,8 +59,8 @@
-  private AtomicBoolean launchedContainer;
-  private AtomicBoolean stoppedContainer;
+  private AtomicBoolean runningPublished;
+  private AtomicBoolean completed;
```
better name it `finished` or `finalStatePublished`? since it's set to true if any of the final states completed, killed or failed is published

chtyim replied:
Updated to `endStatePublished`.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/common/twill/TwillAppLifecycleEventHandler.java:
```diff
@@ -191,15 +188,13 @@ public void containerStopped(String runnableName, int instanceId, String contain
-          programStateWriter.error(programRunId, new RuntimeException(errorMessage));
+        // For workflow, MapReduce, and spark, if there is an error, record the error state
+        if (completed.compareAndSet(false, true)) {
+          programStateWriter.error(programRunId, new Exception(errorMessage));
```
If other program types stop due to an exception, where do we send the error notification?

chtyim replied:
For other program types, as per the current design, they'll never stop, even there is exception raised in container, as the AM would restart it (and eventually aborted the app if it keeps failing such that the time out abort kicks in). See in comment below in the `default:` block.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/StateChangeListener.java:
```diff
@@ -45,31 +46,56 @@ public StateChangeListener(final ProgramRunId programRunId, @Nullable final Stri
   public void killed() {
-    LOG.trace("Program {} killed.", programRunId);
+    LOG.info("Program {} killed.", programRunId);
     programStateWriter.killed(programRunId);
```
If we don't need to guard against multiple updates in this class, it will be good to add the reason in the javadoc.

chtyim replied:
It is done by the program controller.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/StateChangeListener.java:
```diff
@@ -45,31 +46,56 @@ public StateChangeListener(final ProgramRunId programRunId, @Nullable final Stri
+
+  @Override
   public void alive() {
+    LOG.info("Program {} is alive.", programRunId);
```
should this be kept as info?

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/StateChangeListener.java:
```diff
@@ -45,31 +46,56 @@ public StateChangeListener(final ProgramRunId programRunId, @Nullable final Stri
   public void suspended() {
-    LOG.trace("Suspending Program {} .", programRunId);
+    LOG.info("Suspending Program {} .", programRunId);
     programStateWriter.suspend(programRunId);
```
Looks like `StateChangeListener` is only used in local mode, if so where do we publish suspend and resume events in distributed mode?

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -17,86 +17,84 @@
+
   private static final Logger LOG = LoggerFactory.getLogger(AbstractNotificationSubscriberService.class);
   // Sampling log only log once per 10000
   private static final Logger SAMPLING_LOG = Loggers.sampling(LOG, LogSamplers.limitRate(10000));
```
10000 might be too much since the max num retries is only 5000?

chtyim replied:
this is for lifetime.

anew commented:
@chtyim Can you explain what the problem was, and how this fixes it? This is quite a large PR for a bug fix and I am not sure I understand the changes. thanks

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -107,97 +105,148 @@ protected void shutDown() {
+        });
+
+        metricsContext.gauge("process.duration.ms", stopwatch.elapsedTime(TimeUnit.MILLISECONDS));
+        metricsContext.increment("process.notifications", getProcessedCount(messages, messageId));
```
can add a processedCount field in `NotificationIterator` and get it here

chtyim replied:
Updated.

poornachandra reviewed (commented):
@chtyim I did a pass over, and have a couple of question on state publishing.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -107,97 +105,148 @@ protected void shutDown() {
+            processNotifications(context, iterator);
+            String lastMessageId = iterator.getLastMessageId();
+
+            // If the processNotifications hasn't consumed anything, there is nothing to save.
```
have not been consumed

chtyim replied:
Updated.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -107,97 +105,148 @@ protected void shutDown() {
+            String lastMessageId = iterator.getLastMessageId();
+
+            // If the processNotifications hasn't consumed anything, there is nothing to save.
+            if (lastMessageId == null) {
```
can also initialize `NotificationIterator` with `messageId` as `lastMessageId` so that can always return `lastMessageId`

chtyim replied:
That kind of making the contract unclear. The current contract of the `NotificationIterator.getLastMessageId` is to return the last message id of the message being returned by the last call to `next()`. It's better to keep that simple.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -43,173 +54,199 @@
-      // Ignore notifications which specify an invalid ProgramRunId or ProgramRunStatus
+      // Ignore notifications which specify an invalid ProgramRunId or ProgramRunStatus, which shouldn't happen
       if (programRun == null || programStatus == null) {
+        SAMPLING_LOG.warn("Ignore notification that misses program run state information, {}", notification);
```
why `SAMPLING_LOG` if this shouldn't happen?

chtyim replied:
Right. That was from copy and paste.

chtyim commented:
@anew The commits comment pretty much explained it. Here they are:

* Class refactoring/enhancement + bug fixing
   * Allow batch processing of notifications
   * Allow partial processing of a batch of messages
   * Added metrics for progress tracking 
   * Added trace log - Use transaction to fetch data partition events
   * (regression) Use transaction to fetch data partition events
* Correct program state publishing
   * Avoid publishing duplicate completion states from the event handler
   * Implement StateChangeListener.init to publish states in correct order.
* Fix flaky core scheduler test

ecfm commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -2522,6 +2481,88 @@
+
+  <property>
+    <name>system.notification.retry.policy.max.retries</name>
+    <value>5000</value>
```
is this necessary with the following `max.time.secs` specified? It also seems too large

chtyim replied:
The `RetryStrategies` requires it. It is high because in most situation, we don't want it to give up.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -107,97 +105,148 @@ protected void shutDown() {
-        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.transactionalFetch = transactionalFetch;
+      this.emptyFetchDelayMillis = cConf.getLong(Constants.Scheduler.EVENT_POLL_DELAY_MILLIS);
+      this.retryStrategy = RetryStrategies.fromConfiguration(cConf, "system.notification.");
```
I think `ProgramNotificationSubscriberService.java` should have different retry strategies from `ScheduleNotificationSubscriberService.java`, even different topics in `ScheduleNotificationSubscriberService.java` can have different retry strategies.

chtyim replied:
Like how many do we need? And why?

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -215,65 +264,105 @@ public Void call(DatasetContext context) throws Exception {
+      LOG.trace("Fetching system topic '{}' with messageId '{}'", messageId);
+      MessageFetcher messageFetcher = messagingContext.getMessageFetcher();
+      try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                      topic, 100, messageId)) {
```
should the 100 be configurable (like, for example, the sleep time)

chtyim replied:
Will do.

chtyim replied:
fixed

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -215,65 +264,105 @@ public Void call(DatasetContext context) throws Exception {
-     * @param lastFetchedMessageId the message id to persist
-     */
-    public abstract void updateMessageId(DatasetContext context, String lastFetchedMessageId);
+      return new MessageId(Bytes.fromHexString(messageId)).getPublishTimestamp();
```
not sure where this is called... if it is called frequently, it may be better to have a static method in MessageId that extracts only the time publish stamp from the hex string

anew replied:
ah never mind, I reviewed commit by commit and did not see this was already done.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java:
```diff
@@ -215,65 +264,105 @@ public Void call(DatasetContext context) throws Exception {
+      if (messageId == null) {
+        return 0;
+      }
+      // Search backward, since most of the time the whole list is consumed.
```
not sure if that is true if you always fetch 100, some consumers may not want to process that many

anew replied:
I wonder if you can just keep a count while the messages are being processsed

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -43,173 +54,199 @@
       // Check if the program exists
       if (getProgramSpecFromApp(meta.getSpec(), programRunId) == null) {
-        LOG.warn(new NotFoundException(programId).getMessage());
+        LOG.warn("Ignore notification for program {} that doesn't exist, {}", programRunId, notification);
```
Is it possible to remove a program while that program is running? By redeploying the app without that program? Then you would get notifications for that program when it completes.

chtyim replied:
I think we don't allow removing a program (via app redeployment) if the program is still running.

If the app doesn't exists, updating AppMetaStore will fail. Probably better to update anyway via `AppMetaStore`, but catch and ignore the `NotFoundException`, such that there is only one place to handle the not found logic.

chtyim replied:
Just looked at the `AppMetaStore` again, and it actually doesn't check, so the check has to happen here.

anew replied:
ok

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -43,173 +54,199 @@
           if (userArgumentsString == null || systemArgumentsString == null) {
-            LOG.warn((userArgumentsString == null) ? "user" : "system" + " arguments was not specified in program " +
-                     "status notification for program run " + programRunId);
+            LOG.warn("Ignore notification for program {} without {} arguments, {}",
```
Include the "STARTING" state in the message?

chtyim replied:
fixed

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -43,173 +54,199 @@
         case RUNNING:
           long logicalStartTimeSecs = getTimeSeconds(notification.getProperties(),
                                                      ProgramOptionConstants.LOGICAL_START_TIME);
           if (logicalStartTimeSecs == -1) {
```
how can these things happen? Only because of internal errors?

chtyim replied:
yes.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -43,173 +54,199 @@
             return;
           }
           String errorString = properties.get(ProgramOptionConstants.PROGRAM_ERROR);
           BasicThrowable cause = (errorString == null) ? null : GSON.fromJson(errorString, BasicThrowable.class);
```
what happens if fromJson() throws an exception?

chtyim replied:
fixed by catching exception.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -245,5 +286,27 @@ private long getTimeSeconds(Map<String, String> properties, String option) {
+        AppMetadataStore appMetadataStore = new AppMetadataStore(table, cConf, upgradeComplete);
+        // If upgrade was not complete, check if it is and update boolean
+        if (!upgradeComplete.get()) {
+          boolean isUpgradeComplete = appMetadataStore.isUpgradeComplete(APP_VERSION_UPGRADE_KEY);
```
is app version upgrade still needed? Do we support upgrading from such old CDAP versions (pre-3.6)?

chtyim replied:
@maochf @poornachandra do you know? This is from current code, which I just move it from parent class to here.

ecfm replied:
if we don't allow upgrading from pre-3.6 version to 4.3, then it's not needed

chtyim replied:
@maochf how about 3.5 to 4.3?

ecfm replied:
I'm not sure about this...

anew replied:
let's leave it here since it is in existing, working code.

anew reviewed (commented):
mostly minor comments from me.

anew commented:
LGTM

chtyim commented:
Squashed, prepare to merge.

sreevatsanraman commented:
Build passed. Merging.
